tests/mir-opt/casts.rs

```diff
@@ -0,0 +1,11 @@
+#![crate_type = "lib"]
+
+// EMIT_MIR casts.redundant.PreCodegen.after.mir
+pub fn redundant(x: *const u8) -> *const u8 {
+    x as *const u8
+}
```

Comment:
Is this testing anything? Rust playground is telling me that this hasn't changed anything from the MIR that gets emitted for `redundant` on stable.

Reply:
Sure wasn't testing anything. I wish we had a better way to test stuff like this. The easy way out is to look at the pass's diff, but I don't care care very much what this particular pass does, only that the pattern gets optimized away.

Reply:
We should definitely have a `// unit-test` for this case. Specifically, I'd suggest

```rust
// unit-test InstCombine

pub fn redundant_primitive_cast(x: *mut u8) -> *mut u8 {
    x as *mut u8
}

pub fn redundant_cast_with_lifetimes<'a, 'b: 'a>(x: *const &'b u8) -> *const &'a u8 {
    x as *mut u8
}
```

> I don't care care very much what this particular pass does

The goal of this kind of a test is to protect against someone introducing a bug into the pass that causes it not to do anything anymore (accidentally adds an `if false`, case gets forgotten in a refactor, that type of thing). You can additionally have a `casts_e2e.rs` test with all the `PreCodegen.after` examples you'd like.

Reply:
The `InstCombine.diff` for those examples are empty. I feel like there's something in MIR building that is skipping emitting the cast, and the situation I'm trying to optimize here only arises after inlining.